crates/sui-core/src/authority_server.rs

```diff
@@ -269,70 +267,62 @@ impl Validator for ValidatorService {
+        certificate.is_verified = true;
 
-        let tx_digest = transaction.digest();
+        // 2) Check idempotency
+        let digest = certificate.digest();
+        if let Some(response) = self
+            .state
+            .check_tx_already_executed(digest)
+            .await
+            .map_err(|e| tonic::Status::internal(e.to_string()))?
+        {
+            return Ok(tonic::Response::new(response));
+        }
+
+        // 3) If it's a shared object transaction and requires consensus, we need to do so.
+        // This will wait until either timeout or we have heard back from consensus.
+        if certificate.contains_shared_object()
+            && !self
+                .state
+                .transaction_shared_locks_exist(&certificate)
+                .await
+                .map_err(|e| tonic::Status::internal(e.to_string()))?
+        {
+            self.consensus_adapter
+                .submit(&ConsensusTransaction::UserTransaction(Box::new(
+                    certificate.clone(),
+                )))
+                .await
+                .map_err(|e| tonic::Status::internal(e.to_string()))?;
+        }
+
+        // 4) Execute the certificate.
+        let tx_digest = certificate.digest();
         let span = tracing::debug_span!(
-            "process_cert",
+            "execute_transaction",
             ?tx_digest,
-            tx_kind = transaction.data.kind_as_str()
+            tx_kind = certificate.data.kind_as_str()
         );
 
-        let confirmation_transaction = ConfirmationTransaction {
-            certificate: transaction,
-        };
-
-        let info = self
+        let response = self
             .state
-            .handle_confirmation_transaction(confirmation_transaction)
+            .handle_certificate(certificate)
             .instrument(span)
             .await
             .map_err(|e| tonic::Status::internal(e.to_string()))?;
 
-        Ok(tonic::Response::new(info))
-    }
-
-    async fn consensus_transaction(
-        &self,
-        request: tonic::Request<ConsensusTransaction>,
-    ) -> Result<tonic::Response<TransactionInfoResponse>, tonic::Status> {
-        let transaction = request.into_inner();
-        let certificate = match transaction.clone() {
-            ConsensusTransaction::UserTransaction(certificate) => certificate,
-            _ => {
-                let error = SuiError::UnexpectedMessage;
-                return Err(tonic::Status::internal(error.to_string()));
-            }
-        };
-
-        // In some cases we can skip consensus for shared-object transactions: (i) we already executed
-        // the transaction, (ii) we already assigned locks to the transaction but failed to execute it.
-        // The later scenario happens when the authority missed some of the transaction's dependencies;
-        // we can thus try to re-execute it now.
-        let info = match self
-            .state
-            .try_skip_consensus(*certificate)
```

Comment:
So, with this change, even if a cert has been through consensus but not executed (for whatever reason), we will always input it to consensus again before we can execute it. I'm not saying that's not ok, but it's probably the main thing to consider regarding this change.